tx-parser/src/lib.rs

```diff
@@ -30,181 +34,150 @@ fn fn_parser<O, I>(f: fn(I) -> ParseResult<O, I>, err: &'static str) -> TxParser
 
-    fn keyword() -> TxParser<NamespacedKeyword, I> {
-        fn_parser(Tx::<I>::keyword_, "keyword")
-    }
+// Manually expanding the above:
+// def_value_parser_fn!(Tx, integer, i64, input, {
+//     satisfy_map(|x: edn::Value| $transformer(&x)                                      ).parse_stream(input)
+//     satisfy_map(|x: edn::Value| if let Value::Integer(y) = x { Some(y) } else { None }).parse_stream(input)
+// });
 
-    fn keyword_(input: I) -> ParseResult<NamespacedKeyword, I> {
-        return satisfy_map(|x: Value| if let Value::NamespacedKeyword(y) = x {
-                Some(y)
-            } else {
-                None
-            })
-            .parse_stream(input);
-    }
+def_parser_fn!(Tx, keyword, Value, NamespacedKeyword, input, {
+    return satisfy_map(|x: Value| if let Value::NamespacedKeyword(y) = x {
+            Some(y)
+        } else {
+            None
+        })
+        .parse_stream(input);
+});
 
-    fn entid() -> TxParser<Entid, I> {
-        fn_parser(Tx::<I>::entid_, "entid")
-    }
+def_parser_fn!(Tx, entid, Value, Entid, input, {
+    let p = Tx::<I>::integer()
+        .map(|x| Entid::Entid(x))
+        .or(Tx::<I>::keyword().map(|x| Entid::Ident(x)))
+        .parse_lazy(input)
+        .into();
+    return p;
+});
 
-    fn entid_(input: I) -> ParseResult<Entid, I> {
-        let p = Tx::<I>::integer()
-            .map(|x| Entid::Entid(x))
-            .or(Tx::<I>::keyword().map(|x| Entid::Ident(x)))
-            .parse_lazy(input)
-            .into();
-        return p;
-    }
+def_parser_fn!(Tx, lookup_ref, Value, LookupRef, input, {
+    return satisfy_map(|x: Value| if let Value::Vector(y) = x {
+            let mut p = (Tx::<&[Value]>::entid(), any(), eof())
+                .map(|(a, v, _)| LookupRef { a: a, v: v });
+            let r = p.parse_lazy(&y[..]).into();
+            match r {
+                Ok((r, _)) => Some(r),
+                _ => None,
+            }
+        } else {
+            None
+        })
+        .parse_stream(input);
+});
 
-    fn lookup_ref() -> TxParser<LookupRef, I> {
-        fn_parser(Tx::<I>::lookup_ref_, "lookup-ref")
-    }
+def_parser_fn!(Tx, entid_or_lookup_ref, Value, EntidOrLookupRef, input, {
+    let p = Tx::<I>::entid()
+        .map(|x| EntidOrLookupRef::Entid(x))
+        .or(Tx::<I>::lookup_ref().map(|x| EntidOrLookupRef::LookupRef(x)))
+        .parse_lazy(input)
+        .into();
+    return p;
+});
 
-    fn lookup_ref_(input: I) -> ParseResult<LookupRef, I> {
-        return satisfy_map(|x: Value| if let Value::Vector(y) = x {
-                let mut p = (Tx::<&[Value]>::entid(), any(), eof())
-                    .map(|(a, v, _)| LookupRef { a: a, v: v });
-                let r = p.parse_lazy(&y[..]).into();
-                match r {
+// TODO: abstract the "match Vector, parse internal stream" pattern to remove this boilerplate.
+def_parser_fn!(Tx, add, Value, Entity, input, {
+    return satisfy_map(|x: Value| -> Option<Entity> {
```

Comment:
You can omit the `return` and the closing semicolon, saving yourself some space.